lib/plug/builder.ex

```diff
@@ -212,11 +214,24 @@ defmodule Plug.Builder do
       :function -> "expected #{inspect plug}/2 to return a Plug.Conn"
     end
 
+    halt_message = case plug_type do
+      :module   -> "plug pipeline halted in #{inspect plug}.call/2"
+      :function -> "plug pipeline halted in #{inspect plug}/2"
+    end
+
     quote do
       case unquote(compile_guards(call, guards)) do
-        %Plug.Conn{halted: true} = conn -> conn
-        %Plug.Conn{} = conn             -> unquote(acc)
-        _                               -> raise unquote(error_message)
+        %Plug.Conn{halted: true} = conn ->
+          if log_on_halt_level = unquote(builder_opts[:log_on_halt]) do
```

Comment:
You get the warnings because the value of the option is know at compile time! You should have this outside of the quote:

```elixir
if level = build_opts[:log_on_halt] do
  message = ...
  quote do
    require Logger
    Logger.log(unquote(level), unquote(message)
    conn
  end
else
  quote do: conn
end
```